### PR TITLE
Make P2P shuffle extensible

### PR DESCRIPTION
**Pull request by hendrikmakait**
This PR replaces hard-coded logic for handling dataframe shuffles and array rechunking with a dispatch mechanism. The main goal is to improve the separation of concerns and making the P2P mechanism extensible for other kinds of "shuffle-like" operations.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by hendrikmakait** [comment]
The main thing happening in this PR is the introduction of dispatches and replacing some dictionaries with dataclasses. The rest is moving stuff around, which I will probably extract into a separate PR to make it easier to track changes.

**Comment by hendrikmakait** [comment]
@wence-: An initial iteration on this one is done, I would appreciate your thoughts!

**Comment by hendrikmakait** [comment]
To implement a new shuffle-like, you now need to implement...
a few data classes
* subclass of `ShuffleSpec`
* subclass of `ShuffleRunSpec`
* subclass of `SchedulerShuffleState`

a class handling the main logic
* subclass of `ShuffleRun`

a few dispatches to convert things
* `spec_to_scheduler_state`
* `scheduler_state_to_run_spec`
* `run_spec_to_worker_run`

Ideally, I would like to reduce the implementation overhead, but I not sure yet how to best do that.

File: distributed/shuffle/_core.py [hunk 1]
```diff
@@ -275,3 +282,33 @@ def __str__(self) -> str:
...
+    worker_for: dict[_T_partition_id, str]
+    output_workers: set[str]  # TODO: Is this necessary?
+
+
+@dataclass(eq=False)
+class SchedulerShuffleState:
+    _run_id_iterator: ClassVar[itertools.count] = itertools.count(1)
+
+    id: ShuffleId
+    run_id: int
```
Comment (wence-) [inline distributed/shuffle/_core.py:305]: As mentioned below, perhaps:

```suggestion
    id: ShuffleId
    run_id: int = dataclasses.field(init=False, default_factory=partial(next, itertools.count(1))
```
And now subclasses are no longer responsible for ensuring that the sequence number is unique.

File: distributed/shuffle/_rechunk.py [hunk 2]
```diff
@@ -441,47 +447,98 @@ def _() -> dict[str, tuple[NDIndex, bytes]]:
         return self.run_id
 
     async def get_output_partition(
-        self, partition_id: NDIndex, key: str, meta: pd.DataFrame | None = None
+        self, partition_id: NDIndex, key: str, **kwargs: Any
```
Comment (wence-) [inline distributed/shuffle/_rechunk.py:450]: Is this switch to `kwargs` future-proofing?

**Review by wence- — COMMENTED** [review]
A few small suggestions in this round. Perhaps handling the dispatch as required methods on the base dataclass objects is cleaner?

File: distributed/shuffle/_rechunk.py [hunk 3]
```diff
@@ -441,47 +447,98 @@ def _() -> dict[str, tuple[NDIndex, bytes]]:
...
+def _array_run_spec_to_run(
+    spec: ArrayRechunkRunSpec, plugin: ShuffleWorkerPlugin
+) -> ShuffleRun:
+    return ArrayRechunkRun(
+        worker_for=spec.worker_for,
+        output_workers=spec.output_workers,
+        old=spec.old,
+        new=spec.new,
+        id=spec.id,
+        run_id=spec.run_id,
```
Comment (wence-) [inline distributed/shuffle/_rechunk.py:527]: It seems sad that the `ArrayRechunkRun` object loses all the grouping of information (e.g. why do we not just pass the `spec` object in here?)

File: distributed/shuffle/_worker_plugin.py [hunk 4]
```diff
@@ -305,66 +289,11 @@ async def _(
 
                 self.worker._ongoing_background_tasks.call_soon(_, self, existing)
 
-        shuffle = self._create_shuffle_run(shuffle_id, result)
+        shuffle: ShuffleRun = run_spec_to_worker_run(result, self)
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/shuffle/_worker_plugin.py:292]: Agreed, initially I liked the clear separation of having pure dataclasses but I think the dispatch makes it unnecessarily complex.

File: distributed/shuffle/_rechunk.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/shuffle/_rechunk.py:450]: Future-proofing and making it completely transparent to the plugin.

File: distributed/shuffle/_rechunk.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/shuffle/_rechunk.py:527]: I'm currently working on a slightly alternative approach that tries to roundtrip the initial `ShuffleSpec` and avoid some subclassing in the process.

File: distributed/shuffle/_core.py [hunk 5]
```diff
@@ -256,19 +256,60 @@ class ShuffleType(Enum):
...
+        """TODO"""
+
+    @abc.abstractmethod
+    def initialize_run_on_worker(
+        self,
+        run_id: int,
+        worker_for: dict[_T_partition_id, str],
+        plugin: ShuffleWorkerPlugin,
+    ) -> ShuffleRun:
+        """TODO"""
```
Comment (hendrikmakait) [inline distributed/shuffle/_core.py:297]: All you need to do now is implement a concrete subclass of `ShuffleSpec` and `ShuffleRun`.

**Review by wence- — APPROVED** [review]
Thanks, I think this looks neater now. Coverage went made in the tests, but not sure if that will fix itself in the latest test run.

**Pull request merged**